File: Python/CreationalDesignPattern/CompositionDesignPattern.py
from abc import ABC, abstractmethod
from typing import List


class IFileSystemNode(ABC):
    @abstractmethod
    def get_size(self) -> int:
        pass

    @abstractmethod
    def get_name(self) -> str:
        pass

    @abstractmethod
    def tree(self):
        pass


class File(IFileSystemNode):
    def __init__(self, name: str, content: str):
        self.name = name
        self.content = content

    def get_size(self) -> int:
        return len(self.content)

    def get_name(self) -> str:
        return self.name

    def tree(self):
        print(f"File {self.name}")


class Folder(IFileSystemNode):
    def __init__(self, name: str):
        self.name = name
        self.childrens: List[IFileSystemNode] = []

    def touch(self, name: str, content: str):
        self.childrens.append(File(name, content))

    def mkdir(self, name: str):
        self.childrens.append(Folder(name))

    def get_size(self) -> int:
        size = 0
        for node in self.childrens:
            size += node.get_size()
        return size

    def get_name(self) -> str:
        return self.name

    def tree(self):
        print(f"Folder:-> {self.name}")
        for node in self.childrens:
            node.tree()


class FileSystemManager:
    def __init__(self):
        self.path_to_cd = None

    def go_to_current_directory(self):
        pass


def main():
    # Root
    # System.txt
    # Doc
    # pradeep_lld.txt
    # DSA
    # pradeep_dsa.txt

    root = Folder("root")
    root.touch("System.txt", "dasdsadasd")
    root.mkdir("Doc")


if __name__ == "__main__":
    main()

File: C++/CreationalDesignPattern/CompositionDesignPattern.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

struct IFileSystemNode {
    virtual int getSize() = 0;
    virtual std::string getName() = 0;
    virtual void tree() = 0;
    virtual ~IFileSystemNode() = default;
};

class File : public IFileSystemNode {
    std::string name;
    std::string content;

public:
    File(std::string name, std::string content)
        : name(std::move(name)), content(std::move(content)) {}

    int getSize() override { return static_cast<int>(content.length()); }

    std::string getName() override { return name; }

    void tree() override { std::cout << "File " << name << std::endl; }
};

class Folder : public IFileSystemNode {
    std::string name;
    std::vector<std::unique_ptr<IFileSystemNode>> childrens;

public:
    explicit Folder(std::string name) : name(std::move(name)) {}

    void touch(const std::string& name, const std::string& content) {
        childrens.push_back(std::make_unique<File>(name, content));
    }

    void mkdir(const std::string& name) {
        childrens.push_back(std::make_unique<Folder>(name));
    }

    int getSize() override {
        int size = 0;
        for (auto& node : childrens) {
            size += node->getSize();
        }
        return size;
    }

    std::string getName() override { return name; }

    void tree() override {
        std::cout << "Folder:-> " << name << std::endl;
        for (auto& node : childrens) {
            node->tree();
        }
    }
};

class FileSystemManager {
public:
    std::string pathToCd;

    void goToCurrentDirectory() {}
};

int main() {
    // Root
    // System.txt
    // Doc
    // pradeep_lld.txt
    // DSA
    // pradeep_dsa.txt

    Folder root("root");
    root.touch("System.txt", "dasdsadasd");
    root.mkdir("Doc");

    return 0;
}
